src/ape/pytest/contextmanagers.py

```diff
@@ -1,26 +1,141 @@
+import re
+from collections import deque
 from typing import Optional, Type
 
-from ape.exceptions import ContractLogicError, TransactionError
+import ape
+from ape.exceptions import APINotImplementedError, ContractLogicError, TransactionError
 
 
 class RevertsContextManager:
-    def __init__(self, expected_message: Optional[str] = None):
+    def __init__(self, expected_message: Optional[str] = None, dev_message: Optional[str] = None):
+        if dev_message is not None and expected_message is not None:
+            raise ValueError("Cannot supply both 'expected_message' and 'dev_message'.")
+
         self.expected_message = expected_message
+        self.dev_message = dev_message
 
-    def __enter__(self):
-        pass
+    def _check_dev_message(self, exception: ContractLogicError):
+        """
+        Attempts to extract a dev-message from the contract source code by inspecting what
+        instruction(s) led to a transaction revert.
 
-    def __exit__(self, exc_type: Type, exc_value: Exception, traceback) -> bool:
-        if exc_type is None:
-            raise AssertionError("Transaction did not revert.")
+        Raises:
+            AssertionError: When the trace or source can not be retrieved, the trace does not end in
+            a REVERT instruction, or the dev message cannot be found or does not match the expected
+            dev message.
+        """
+        if exception.txn is None:
+            raise AssertionError("Could not fetch transaction information to check dev message.")
 
-        if not isinstance(exc_value, ContractLogicError):
+        receipt = exception.txn.provider.get_receipt(
+            exception.txn.txn_hash,
+            raise_on_fail=False,
+        )
+
+        try:
+            contract = ape.Contract(receipt.receiver)
```

Comment:
`receipt.receiver` is optional I think, we may need to handle that.

Reply:
Updated in force push to 782e88fe